### PR TITLE
Disable Materializing Grads

### PR DESCRIPTION
**Pull request by iK1D**
Disable materializing grads in forward so that None object will not be converted to a tensor filled with zeros prior to calling backward. We will materialize the output grads by ourselves. If the output grad is None when calling backward, when it's the direct input of backward graph, we will materialize it to all-0 tensor with same shape of output, otherwise, we will use the scalar-0 tensor, because in this case there will be Add node added, and the scalar-0 tensor is always OK as one of inputs of the Add node.

This PR also contains a fix for module gradient builder. With the fix, we will not ignore any output grads from PT, if it's output of one node, we will add an Add node to add the node output with the grad tensor from PT.

Take the BERT large as example, change the user outputs to return all of loss, prediction_scores and seq_relationship_score, but call the backward from loss only. Without set_materialize_grads(False), the max batch size is 80 (on 32G V100), when set to False, the max batch size can reach to 88. Python debugger also confirms that when set to False, the output grads for prediction_scores and seq_relationship_score are None instead of all-zero tensors.

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
It's great that this worked!

**Review by SherlockNoMad — DISMISSED** [review]

**Comment by mrry** [comment]
This is a really nice saving, but can you explain what happens in the following corner case, starting from your example:

> Take the BERT large as example, change the user outputs to return all of loss, prediction_scores and seq_relationship_score

Now we call `backward()` on a tensor that's derived from `loss` and `prediction_scores`, e.g. with this silly example:

```python
loss, prediction_scores, _ = model.forward()
new_loss = loss + prediction_scores.sum()
new_loss.backward()
```

Do we do the right thing in this case? It doesn't look like this should change the control flow through `ORTModule` compared to calling `loss.backward()`, so I'd be surprised if it worked without more changes....

**Comment by iK1D** [comment]
@mrry Thanks for your example. Actually for this case, both loss_grad and prediction_scores_grad will be all-1 tensors, and seq_relationship_score_grad will be None from PY. Set the flag to False will keep seq_relationship_score_grad as None instead of changing it to all-0 tensor, and it will work as current implementation needs loss_grad only.
Actually this reveal another big issue in our gradient graph builder, our current implementation is wrong for this case. Here since prediction_scores_grad is an output of a node in the graph, so during the graph builder, we ignore the all-1 prediction_scores_grad from PT. The right behavior should be to add a Sum node to sum the node output and the all-1 tensor from PT as the final prediction_scores_grad.
I modified the graph builder by adding the Sum node, but seems there are still some other issue during Yield op execution so the calculation result still not correct. I will go on debug this.

**Comment by mrry** [comment]
Thanks for digging into this @iK1D! There are probably some interesting complexity-efficiency tradeoffs to be made here, for example:

1. (Easiest, least efficient) We could be pessimistic and always assume that *any* of the forward outputs could have a gradient, but I think that would require us to materialize the grads in every case, and thus we'd waste memory and compute.
2. (Harder, more efficient) Since the set of forward outputs that have gradients is probably pretty stable from one run to the next, we could optimistically specialize the graph for the common case, and provide some way to fall back to (1) if we get unexpected gradients.
3. (Hardest, most(?) efficient, and probably needs more thought!) We could disable materializing grads, as you have in this PR, and put some `If` nodes into the graph to deal with the `None` case. For example, each backward input could become a tuple of (bool, value), or some other representation that covers the fact that they are "optional". Then we could wrap the backwards computation for each input in the true branch of an `If` block, with a passthrough in the false branch. Depending on how efficient the control-flow implementation is, we might still want to specialize as in (2). 

I'd prefer us to implement (1) in the first place, because we can always specialize the code manually by wrapping the original module in another module that drops the irrelevant outputs. 

There are probably other approaches that could work here as well, but wanted to share a brain dump....

**Comment by SherlockNoMad** [comment]
@mrry, The auxiliary loss in the MoE model is actually constructed in a similar way you shown above, so this is not an uncommon case. Sample here: https://aiinfra.visualstudio.com/Lotus/_git/MOE?path=%2Fmoe_module%2Fmoe.py&version=GBmain&line=33&lineEnd=34&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents 

@iK1D, indeed there is a bug on how we currently construct the gradient graph. 

Since we can't tell which module outputs would end up getting the gradient, I think we have two options: 
1) Always assume all outputs has gradients, and trim the graph in the first backward call(). 
2) Defer building training graph to the backward() call 

+@satyajandhyala for more perspectives, since he has also looked at this problem before.

**Comment by iK1D** [comment]
I am thinking Derek's #1, but with some trick. We still set the materialize_grads to False, if it's None, we create a 0-scalar tensor as its grad and pass to ORT. If we can somehow make the SetOutputMLValue work (as the shape is inconsistent), I think the Sum is quick as we have scalar optimization for calculation.

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 1]
```diff
@@ -2220,27 +2220,16 @@ Return true if all elements are true and false otherwise.
...
-          auto typeProto = ctx.getInputType(j);
+        ORT_ENFORCE(ctx.getNumInputs() == ctx.getNumOutputs());
+        for (size_t i = 0; i < ctx.getNumInputs(); ++i) {
+          propagateElemTypeFromInputToOutput(ctx, i, i);
+          auto typeProto = ctx.getInputType(i);
           if (!hasShape(*typeProto)) {
             continue;
           }
-          propagateShapeFromInputToOutput(ctx, j, i);
+          propagateShapeFromInputToOutput(ctx, i, i);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2232]: shape can be different, right?  for scaler 0 ?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 2]
```diff
@@ -134,13 +134,18 @@ def forward(ctx, *inputs, **kwargs):
...
+                    ctx.set_materialize_grads(False)
+
                     # Use IO binding
                     _create_iobinding(self._training_io_binding, inputs, self._onnx_training, self._device)
 
                     # Run and return module outputs.
                     forward_outputs, run_id = self._training_session.run_forward(self._training_io_binding, self._run_options)
                     user_outputs = tuple(_ort_output_to_torch_tensor(forward_output) for forward_output in forward_outputs)
                     ctx.run_id = run_id
+                    # Save shape, device and type info for materializing tensor in backward if output grad is None.
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule.py:147]: nit. move to the line after "ctx.set_materialize_grads(False)", since they are connected.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 3]
```diff
@@ -152,14 +157,17 @@ def backward(ctx, *grad_outputs):
...
-                    # we filter out the invalid entries in grad_outputs, accessing using the mapped index.
-
-                    for _, i in self._onnx_graphs_info.backward_output_grad_names_map.items():
-                        grad_output = grad_outputs[i]
-                        if not grad_output.is_contiguous():
+                    contiguous_grad_outputs = []
+                    for idx, grad_output in enumerate(grad_outputs):
+                        if grad_output is None:
+                            if idx in self._onnx_graphs_info.output_grad_indices_require_materialized:
+                                grad_output = torch.zeros(ctx.output_info[idx][0], device=ctx.output_info[idx][1], dtype=ctx.output_info[idx][2])
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule.py:164]: nit, for readability 
shape, device, type = ctx.output_info[idx]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -614,22 +614,43 @@ def test_input_requires_grad_backward_creates_input_grad_as_required0(device):
 
     _, y2 = model(x1,x2)
     s2 = y2.sum()
-    s2.backward()
+    s2.backward()   # y1's gradient will be materialized.
     named_params = list(model.named_parameters())
     assert named_params[0][1].grad is None or torch.count_nonzero(named_params[0][1].grad) == 0
     assert named_params[1][1].grad is None or torch.count_nonzero(named_params[1][1].grad) == 0
     assert torch.count_nonzero(named_params[2][1].grad) > 0
     assert torch.count_nonzero(named_params[3][1].grad) > 0
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:622]: let's also compare the model's gradient result with pytorch ? just like what you did in the test below test_loss_combines_two_outputs_with_dependency

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 5]
```diff
@@ -614,22 +614,43 @@ def test_input_requires_grad_backward_creates_input_grad_as_required0(device):
...
+@pytest.mark.parametrize("device", ['cuda'])
+def test_loss_combines_two_outputs_with_dependency(device):
+
+    def run_step(model, x1, x2):
+        y1, y2 = model(x1, x2)
+        loss = y1.sum() + y2.sum()
+        loss.backward()
+
+    N, D_in, H, D_out = 32, 784, 500, 10
+    pt_model = NeuralNetMultiplePositionalArgumentsMultipleOutputs1(D_in, H, D_out).to(device)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:633]: nit. better name, ...MultiOutputsWithDependency

Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:633]: same applies to NeuralNetMultiplePositionalArgumentsMultipleOutputs0 
...MultiOutputsWithoutDependency

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.h [hunk 6]
```diff
@@ -41,9 +41,9 @@ struct TrainingGraphInfo {
   std::vector<std::string> initializer_grad_names_to_train{};
   // The user outputs.
   std::vector<std::string> user_output_names{};
-  // The user output grad names that are actually required by the backward graph 
-  // mapped to the index of the correspoinding output of inference graph.
-  std::unordered_map<std::string, size_t> backward_output_grad_names_map{};
+  // Indices of output grads that need to be materialized to full size all-0 tensor.
+  // Otherwise, we can use scalar-0 tensor.
+  std::vector<size_t> output_grad_indices_require_materialized{};
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/module_gradient_graph_builder.h:46]: nit. output_grad_indices_require_full_shape? 
setting scale 0 is also a form of materialization...

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.cc [hunk 7]
```diff
@@ -179,46 +179,58 @@ void ModuleGradientGraphBuilder::AddYieldOp() {
...
-  for (std::size_t i = 0; i < training_graph_info_.user_output_names.size(); ++i) {
-    const auto& name = training_graph_info_.user_output_names[i];
-    std::string grad_name = name + "_grad";
-    if (non_backward_user_output_grad_names.find(grad_name) == non_backward_user_output_grad_names.end()) {
-      training_graph_info_.backward_output_grad_names_map.insert(std::make_pair(grad_name, i));
-      required_grad.add_ints(static_cast<int64_t>(i));
+  for (const auto& output_grad_name : internal_output_grad_names) {
+    Node* producer_node = gradient_graph.GetMutableProducerNode(output_grad_name);
+    NodeArg* arg = gradient_graph.GetNodeArg(output_grad_name);
+    int producer_node_arg_index = 0;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/module_gradient_graph_builder.cc:196]: In graph_utils.h, there is a GetNodeInputIndexFromInputName

maybe add a GetNodeOutputIndexFromOutputName ?

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.cc [hunk 8]
```diff
@@ -179,46 +179,58 @@ void ModuleGradientGraphBuilder::AddYieldOp() {
...
+    while (producer_node_arg_index < static_cast<int>(producer_node->OutputDefs().size()) &&
+           producer_node->OutputDefs()[producer_node_arg_index] != arg) {
+      producer_node_arg_index++;
     }
+
+    const TypeProto* type_info = producer_node->MutableOutputDefs()[producer_node_arg_index]->TypeAsProto();
+    auto& external_node_arg = gradient_graph.GetOrCreateNodeArg(
+        gradient_graph.GenerateNodeArgName(output_grad_name + "_external"), type_info);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/module_gradient_graph_builder.cc:204]: output_grad_name + "_external"  now becomes reserved pattern
let's make a util function for this, just like the GradientName() ?

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.cc [hunk 9]
```diff
@@ -179,46 +179,58 @@ void ModuleGradientGraphBuilder::AddYieldOp() {
...
 
   std::vector<NodeArg*> yield_input_node_args;
   std::vector<NodeArg*> yield_output_node_args;
-  for (const auto& name : training_graph_info_.user_output_names) {
+  training_graph_info_.output_grad_indices_require_materialized.clear();
+  for (size_t i = 0; i < training_graph_info_.user_output_names.size(); i++) {
+    std::string name = training_graph_info_.user_output_names[i];
     yield_input_node_args.emplace_back(gradient_graph.GetNodeArg(name));
-  }
+    std::string grad_name = name + "_grad";
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/module_gradient_graph_builder.cc:219]: use GradientName

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]
overall, it looks good to me. Thanks a lot for the fix!

**Comment by SherlockNoMad** [comment]
with this PR, we should reach a good assumption
1) Yield's input and output is always 1 to 1 mapping
2) we no longer need to reorder the graph outputs, so Yield's output order should match module's forward output order

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2232]: Scalar-0 is just a workaround during execution. Logically the shape is same. Setting this here can help to infer other shape info in the backward graph.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 2 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule.py:147]: user_outputs are generated after run_forward. Let me move the set_materialize_grads here.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2232]: Aishwarya found an issue in T5 that mem_plan reuse this scalar tensor. I think mem_plan should have some bug, but here I will not set the shape for outputs that might be scalar. Will check the mem_plan issue later.

File: orttraining/orttraining/test/python/_test_helpers.py [hunk 10]
```diff
@@ -156,7 +158,10 @@ def assert_gradients_match_and_reset_gradient(ort_model, pt_model, reset_gradien
         pt_name, pt_param = pt_named_param
 
         assert pt_name in ort_name
-        assert torch.allclose(ort_param.grad, pt_param.grad, rtol=rtol, atol=atol)
+        if pt_name in none_pt_params:
+            assert pt_param.grad is None
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/_test_helpers.py:162]: nit. shall we also assert ort_param.grad as zeros?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 11]
```diff
@@ -654,36 +654,63 @@ def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder
...
+
+    def run_step0(model, x1, x2):
+        y1, _ = model(x1, x2)
+        s1 = y1.sum()
+        s1.backward()   # y2's gradient will be materialized to full shape.
+
+    run_step0(pt_model, pt_x1, pt_x2)
+    run_step0(ort_model, ort_x1, ort_x2)
+
+    assert torch.allclose(ort_x1.grad, pt_x1.grad)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:674]: let's also assert on the forward results? 
I caught a mismatch on forward result in test_input_requires_grad_backward_creates_input_grad_as_required1
see 
https://msdata.visualstudio.com/Vienna/_sprints/taskboard/ONNX%20Training/Vienna/Cobalt?workitem=1064208

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 12]
```diff
@@ -654,36 +654,63 @@ def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder
...
+    pt_x1 = torch.randn(N, D_in, device=device, requires_grad=False)
+    pt_x2 = torch.randn(N, D_in, device=device, requires_grad=False)
+    ort_x1 = pt_x1.clone()
+    ort_x2 = pt_x2.clone()
+
+    # Both y1 and y2's gradients are not None.
+    run_step(pt_model, pt_x1, pt_x2)
+    run_step(ort_model, ort_x1, ort_x2)
+
+    _test_helpers.assert_gradients_match_and_reset_gradient(ort_model, pt_model)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:713]: also assert forward results?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 13]
```diff
@@ -2220,27 +2220,26 @@ Return true if all elements are true and false otherwise.
       .Output(0, "outputs_grad", "Gradient of outputs returned from pytorch.", "T", OpSchema::Variadic,
               /*is_homogeneous*/ false,
               /*min_arity*/ 1)
-      .Attr(
-        "required_grad",
-        "The indices of the outputs that require gradient outputs.",
-        AttributeProto::INTS)
+      .Attr("full_shape_outputs", "The indices of the outputs that must have full shape.", AttributeProto::INTS)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2223]: also add an assert in yield's kernel?
output shape should match input shape if "full_shape" is true.

**Review by SherlockNoMad — APPROVED** [review]
Remaining comments are non blocking, and they are mostly for defensive programming practice.
